src/modules/auth/hooks/useAuth.ts

```diff
@@ -0,0 +1,68 @@
+import React, { useEffect } from 'react';
+import { selectAuthState, selectAuthData, selectCurrentUser } from '../';
+import { useRouter } from 'next/router';
+
+import { useSelector, useDispatch } from 'react-redux';
+import { AuthState, SESSION_KEY, resolveAuthState } from '../redux';
+
+// to be used in app
+export const useAuth = () => {
+  const state = useSelector(selectAuthState), data = useSelector(selectAuthData), user = useSelector(selectCurrentUser);
+
+  return {
+    state, 
+    data, 
+    user, 
+    isResolved: state !== AuthState.Prepare, 
+    isAuthroized: state === AuthState.Authorized,
+    // be care that !isAuthroized !== Unauthorized
+    isUnAuthroized: state === AuthState.Unauthorized,
+  };
+}
+
+// to be used in app
+export const useAuthEffect = () => {
+  const { state, data, user, isResolved, isAuthroized, isUnAuthroized } = useAuth();
+
+  usePersistAuthEffect(data);
+  useRedirectAuthEffect(isResolved, isAuthroized);
+};
+
+export const useRedirectAuthEffect = (isResolved: boolean, isAuthroized: boolean, redirect: boolean = true) => {
+  const router = useRouter();
+  const isAuthPage = router.pathname.includes('/auth/');
+
+  useEffect(() => {
+    if(isResolved && redirect) {
+      //redirect to Home if authed & visiting auth page
+      if(isAuthroized && isAuthPage) 
+        router.push('/');
+
+      //redirect to Login if unauth & not visiting auth page
+      if(!isAuthroized && !isAuthPage) 
+        router.push('/auth/signin');
```

Comment:
seems not handle 404 page?

Reply:
since we do not have any "public" page right now, we do not have to put whether it's 404 or not before whether it's authorized or not.
That means we could prompt to login even it's a 404 page.